### ISSUE TITLE
[Inquiry] Missing details about local resource redirection/injection?

### ISSUE THREAD

**Issue by RoxKilly** [gorhill/uBlock#1435]
I have two questions:

In #897, concerning redirecting to a local resource to substitute for a doubleclick script,  @gorhill  wrote: 
> uBO in medium mode would cause static.doubleclick.net/instream/ad_status.js to be blocked regardless of the [3rd-party filter]. The redirection filter will help in such case

That issue concerns a webpage's attempt to use a doubleclick hosted script in order to determine whether the user has adblocking. If doubleclick doesn't respond, the page could conclude that adblocking is on. The 3rd-party filter would thus whitelist this connection to make the page think that adblocking is off. Problem is that this exposed the user to doubleclick. @gorhill was discussing having a redirection filter respond so that the page would think doubleclick was reached (and thus that there is no adblocking) even though no connection was ever made.  This brings me to my first question:

1.  **Do redirection filters take precedence over all others (url filtering, dynamic filtering, static filtering)?**  If I use dynamic filtering to block connections to doubleclick, would the redirection filter still trigger when a page tries to connect to doubleclick?  I've looked in the wiki about [how filtering works](https://github.com/gorhill/uBlock/wiki/Overview-of-uBlock's-network-filtering-engine) with no luck.

2.  The second question is much simpler.  When the `script:inject` filter is used to insert local resources into a page, when is the injected script executed? Before the DOM is parsed? Before any other script? The [docs](https://github.com/gorhill/uBlock/wiki/Static-filter-syntax#special-cosmetic-filters) don't answer this either.

**Comment by gorhill** [comment]
Redirection filters can only be implemented through static filters. Still, a static filter with a `redirect` option is interpreted as two distinct entities: one which is a standard plain static filter, and another one which is an independent redirection directive.

The redirection directive kicks in whenever a matching resource is blocked, regardless of how it is blocked, it can be through a static filter, a dynamic filtering rule, or a URL filtering rule.

Re. `script:inject`: the moment at which the injected script execute is platform dependent. Currently it's done at the same time the specific cosmetic filters are injected, and from what I have observed it's some time after the DOM parsing has started (somewhere while the content of the `head` tag is being parsed).

**Comment by RoxKilly** [comment]
@gorhill sorry but your explanation to the 1st question is not clear. You wrote:
> The redirection directive kicks in whenever a matching resource is blocked, regardless of how it is blocked...

In uBO's privacy filters, I see `||google-analytics.com/analytics.js$script,redirect=google-analytics.com/analytics.js`.  From your explanation, which of the following logic paths is correct?

1.  If there is a block upstream (eg: dynamic filter: `* google-analytics.com * block`) then...
1a.  uBO never sees the redirect because static filters get skipped
1b.  uBO executes the redirect

2.  If there is no block upstream and no other applicable static filter then...
2a.  uBO executes the block and redirect from the privacy filters
2b.  uBO does not execute the redirect

3.  If there is no block upstream but there is another 3rd-party filter which whitelists this script with the `important` tag...
3a.  uBO executes the 3rd-party static filter and allows the connection to google-analytics
3b.  uBO executes the block and redirect and ignores the 3rd-party static filter

Please clarify which of the above is accurate.

**Comment by gorhill** [comment]
https://github.com/gorhill/uBlock/blob/master/src/js/traffic.js#L115

**Comment by RoxKilly** [comment]
My findings (in response to the questions I raised two posts ago):

1.  uBO still executes the redirect even though the block occured further upstream (eg: in dynamic filters).

2.  uBO still executes the redirect even though there is no block upstream, since the redirect rule is itself a block rule (eg taken from uBO's *Privacy* filters: `||googletagservices.com/tag/js/gpt.js$script,redirect=googletagservices.com/gpt.js` blocks first, then redirects)

3.  In the presence of a conflicting static allow filter, even if it is marked important (eg: `@@||googletagservices.com/tag/js/gpt.js$script,important`), uBO still executes the block&redirect. The order in which the filters appear doesn't seem to matter (allow filter first, block&redirect later, or vice-versa)

In short, the block&redirect is considered a static filter that overrides other static filters (from *3rd-party filters* or *My filters* tab) . Unless the resource has a superseding `allow` filter (URL filters and dynamic filters supersede static filters - [see the docs](https://github.com/gorhill/uBlock/wiki/Overview-of-uBlock's-network-filtering-engine)), the block&redirect will be executed.

**Comment by RoxKilly** [comment]
ok I'll rephrase:
3.  In the presence of a conflicting static allow filter such as `@@||googletagservices.com/tag/js/gpt.js$script`, uBO still executes the block&redirect whether the conflicting static filter occurs before or after the block&redirect rule.

**Comment by gorhill** [comment]
> In the presence of a conflicting static allow filter such as `@@||googletagservices.com/tag/js/gpt.js$script`, uBO still executes the block&redirect 

No.

**Comment by RoxKilly** [comment]
Ok.  How come if I add the filter `@@||example.com^$document`, the block&redirect rule `||example.com/script.js$script,redirect=example.js` still gets executed? Isn't that a case in which we have a matching exception filter that conflicts with the block filter for the same resource? Doesn't $document matched every resource at example.com? From the [Adblock filters docs](https://adblockplus.org/filters):
>  If you specify $document option you will get an exception for the entire page

Anyway, thanks for explaining. It has been helpful.

**Comment by gorhill** [comment]
> `@@||example.com^$document`

<https://github.com/gorhill/uBlock/wiki/Static-filter-syntax#not-supported>

**Comment by RoxKilly** [comment]
Aaah it all comes together. I saw `$document` in uBO filters and assume it was supported so that's what I used to run my conflicting rules tests to question 3 and that's why my results were false.
